### PR TITLE
feat(xfn): Consider composite ready state in function response

### PR DESCRIPTION
**Pull request by MisterMX**
### Description of your changes

Marks the composite as ready or unready if the if `.Desired.Composite.Ready` field is set explicitly.

Fixes https://github.com/crossplane/crossplane/issues/6020

I have: 

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `earthly +reviewable` to ensure this PR is ready for review.
- ~[ ] Added or updated unit tests.~
- ~[ ] Added or updated e2e tests.~
- ~[ ] Linked a PR or a [docs tracking issue] to [document this change].~
- [x] Added `backport release-x.y` labels to auto-backport this PR.

Need help with this checklist? See the [cheat sheet].

[contribution process]: https://github.com/crossplane/crossplane/tree/main/contributing
[docs tracking issue]: https://github.com/crossplane/docs/issues/new
[document this change]: https://docs.crossplane.io/contribute/contribute
[cheat sheet]: https://github.com/crossplane/crossplane/tree/main/contributing#checklist-cheat-sheet

### CODE REVIEW

File: apis/apiextensions/fn/proto/v1/run_function.proto [hunk 1]
```diff
@@ -223,9 +223,6 @@ message Resource {
   //
   // * A Function should set this field to READY_TRUE in a RunFunctionResponse
   //   to indicate that a desired composed resource is ready.
-  //
```
Comment (negz) [inline apis/apiextensions/fn/proto/v1/run_function.proto:226]: If we make this change we should document (here and elsewhere) the semantics. Which I think would be:

* If you omit this field, we'll derive XR readiness from composed resource readiness
* If you specify this field, we'll ignore composed resource readiness and use what you explicitly told us

File: cluster/crds/pkg.crossplane.io_deploymentruntimeconfigs.yaml [hunk 2]
```diff
@@ -4312,8 +4313,6 @@ spec:
                                         More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
                                         TODO: Drop `kubebuilder:default` when controller-gen doesn't need it https://github.com/kubernetes-sigs/kubebuilder/issues/3896.
                                       type: string
-                                  required:
-                                  - name
```
Comment (negz) [inline cluster/crds/pkg.crossplane.io_deploymentruntimeconfigs.yaml:4316]: Is this related to the rest of the PR? If not, could you do it in a separate commit?

File: cluster/meta/meta.pkg.crossplane.io_functions.yaml [hunk 3]
```diff
@@ -14,7 +14,7 @@ spec:
     singular: function
   scope: Namespaced
   versions:
-  - name: v1beta1
+  - name: v1
```
Comment (negz) [inline cluster/meta/meta.pkg.crossplane.io_functions.yaml:17]: This also seems unrelated to the rest of the PR. Please do it in a separate commit.

File: internal/controller/apiextensions/composite/reconciler.go [hunk 4]
```diff
@@ -158,6 +158,11 @@ type CompositionResult struct {
 	ConnectionDetails managed.ConnectionDetails
 	Events            []TargetedEvent
 	Conditions        []TargetedCondition
+
+	// ConditionsOverride allows the composition process to force
+	// a specific condition status, unlike Conditions which does not allow
+	// setting system conditions.
+	ConditionsOverride []xpv1.Condition
```
Comment (negz) [inline internal/controller/apiextensions/composite/reconciler.go:165]: I wonder if we could use a special `TargetdCondition` variant. i.e. Add a `System` field and just use that internally to surface these conditions.

Comment (MisterMX) [reply internal/controller/apiextensions/composite/reconciler.go:165]: I tried this but it does work as intended since the custom conditions update (which uses the `Conditions` array) is performed in `handleCommonCompositionResult()` but the system conditions are set in `updateXRConditions()`.

File: cluster/crds/pkg.crossplane.io_deploymentruntimeconfigs.yaml (hunk 2 above)
Comment (MisterMX) [reply cluster/crds/pkg.crossplane.io_deploymentruntimeconfigs.yaml]: No, I think I messed something up with the generator. I removed it.

File: cluster/meta/meta.pkg.crossplane.io_functions.yaml (hunk 3 above)
Comment (MisterMX) [reply cluster/meta/meta.pkg.crossplane.io_functions.yaml]: No, I think I messed something up with the generator. I removed it.

File: internal/controller/apiextensions/composite/reconciler.go (hunk 4 above)
Comment (MisterMX) [reply internal/controller/apiextensions/composite/reconciler.go]: I changed to a new type `CompositeResource` that is analog to `ComposedResource` and contains a `Ready *bool` property.

**Review by phisco — APPROVED** [review]
overal LGTM, code coverage is complaining, but not too much so we could be fine... Thanks @MisterMX! 

@negz I'll leave to you the final call.

**Review by negz — APPROVED** [review]
